cupy/core/_routines_indexing.pyx

```diff
@@ -350,18 +355,12 @@ cdef ndarray _simple_getitem(ndarray a, list slice_list):
     return v
 
 
-_nonzero_kernel_1d = ElementwiseKernel(
```

Comment:
Why did you remove this kernel? Please compare the performance for 1-dim array inputs.

Reply:
Okay, I did a quick benchmark for 1d array and the output is as below,
```python
>>>a = cupy.random.rand(1000000)
>>>print(repeat(cupy.argwhere, (a,)))
argwhere            :   324.478 us   +/-46.795 (min:  291.071 / max: 2166.469) us    414.731 us   +/-46.216 (min:  382.560 / max: 2251.872) us
```

Reply:
@rushabh-v 
1. You should test `nonzero`, not `argwhere`, because you deleted the kernel and replaced with a new implementation.
2. You should also compare with the performance before this PR, and see if your changes improve it or not.

Reply:
We were thinking this was redundant in the `argwhere` case as the result is always a 2-D array and it seemed like we were running into some test failures with it. Did not realize there was a notable performance increase in performance for that kernel. If that's true, agree we should add it back

To Leo's points above, would add one more:

3. Include more `0`s in the data. Maybe `a = (cupy.random.rand(1000000) < 0.5)`?

Reply:
Oh I see, sorry about the noise. Didn't follow the discussion from start. But it'd still be good to have benchmarks done.

Reply:
Okay, so the performance on this patch is:
```python
>>>a = cupy.random.rand(1000000)
>>>print(repeat(cupy.nonzero, (a < 0.5,)))
nonzero             :   419.186 us   +/-23.982 (min:  401.881 / max:  960.754) us    549.686 us   +/-22.333 (min:  530.432 / max: 1090.560) us
```

And before this PR it was,
```python
>>>a = cupy.random.rand(1000000)
>>>print(repeat(cupy.nonzero, (a < 0.5,)))
nonzero             :   282.070 us   +/-46.222 (min:  251.944 / max: 1066.658) us    362.395 us   +/-45.803 (min:  333.696 / max: 1143.040) us
```

Reply:
Ah ok! So it sounds like we do need the 1-D case for performance reasons.

Reply:
Oh yes, I'll be adding that back soon.

Reply:
Sorry for the back-and-forth on this 😞

Reply:
No worries : )
Performance is the most important thing.

Reply:
I have added `_nonzero_kernel_1d` back. And it is passing all the tests.
But I don't know why while running `cupyx.time.repeat` on it for 1d array, it raises,
```python
CompileException: /content/cupy/cupy/core/include/cupy/carray.cuh(285): warning: statement is unreachable
          detected during instantiation of "void CIndexer<_ndim>::set(ptrdiff_t) [with _ndim=1]" 
/tmp/tmpt6doxcch/ecc71d65296e1d43cf7830da23d75472_2.cubin.cu(12): here

/tmp/tmpt6doxcch/ecc71d65296e1d43cf7830da23d75472_2.cubin.cu(15): error: expression must have pointer-to-object type

1 error detected in the compilation of "/tmp/tmpt6doxcch/ecc71d65296e1d43cf7830da23d75472_2.cubin.cu".
```

Reply:
(Note: I was running it on colab.)

Reply:
Oh figured it out,
It's kind of an `IndexError`.

```python
a = cupy.random.rand(1000000)
print(repeat(cupy.nonzero, (a<0.5,))) # raises error
print(repeat(cupy.nonzero, ((a<0.5).reshape(1, 1000000),))) # does not raise error
```

Reply:
Okay, So I have added the `_nonzero_kernel_1d` back. And it's passing all the tests.

And I also did the benchmark on it, and it's output is as follows,
```python
>>>a = cupy.random.rand(1000000)
>>>print(repeat(cupy.nonzero, (a<0.5,)))
nonzero             :   237.272 us   +/-52.330 (min:  207.846 / max: 2926.066) us    269.696 us   +/-51.127 (min:  237.248 / max: 2938.528) us
```

Reply:
Thanks for the update Rushabh! 😄

So it looks like with these changes we *may* even be seeing a speedup, but it's difficult to say with the uncertainty in the measurement. Does that sound right?

Reply:
Actually, I am testing it on Google Colab(by building a development-env there as we do locally). So, if I am not wrong, it might be an uncertainty from the cloud side. Does that make sense?

Reply:
should do the benchmark once again?

Reply:
If you have IPython or Jupyter, you could use `%timeit`, which would be good for measuring this sort of thing.

Reply:
No actually I don't have Nvidia GPU. That's why I am using colab for testing and for doing the benchmarks.

Reply:
Right I think Colab has those things though. Or at least their implementation of a Jupyter Notebook. Would give it a try and see if it works 😉

Reply:
> If you have IPython or Jupyter, you could use %timeit, which would be good for measuring this sort of thing.

It looks like he has been using `cupyx.time.repeat` above? That function forces device synchronization to avoid artificially low timings that can occur without it. So, I wouldn't recommend switching from `repeat` to `%timeit`.

I think @rushabh-v was indicating that there may be uncertainty on the hardware side. I don't think Colab will necessarily have assigned identical hardware if those timings were run across different sessions.

Reply:
Please don't do `%timeit` for any GPU code unless you're certain a synchronization will be done at the end of your code snippet. This is one of the most common pitfalls for pure Python/IPython/Jupyter users.

Reply:
Let me benchmark locally and see if I can get consistent results.

Reply:
Thanks for the feedback Gregory! 😄

Good to know. Wouldn't adding `cupy.cuda.Stream().synchronize()` be sufficient? Or is there more that `repeat` is doing.

The main thing `%timeit` would help with is providing options to gather a sufficient number of runs to really cutdown on standard deviation. Does `repeat` have an option for this?

Should add I don't know that much about the setup on Colab. So may be missing other concerns with benchmarking there.

Reply:
> The main thing %timeit would help with is providing options to gather a sufficient number of runs to really cutdown on standard deviation. Does repeat have an option for this?

Here is the function signature: https://github.com/cupy/cupy/blob/d58bbb0db60dfed1614e6e173927f0fd05f292ee/cupyx/time.py#L43

The first `n_warmup` runs are not included in the reported time and then `n` repetitions are measured. GPU times are recorded using CUDA stream events, but I think that is less important than making sure synchronization is performed.

The other advantage over `%timeit` is that the first set of numbers in the printouts gives indicates the time spent on the CPU side while the second, longer set of numbers includes the GPU kernel time as well. I think the output could be better labeled to indicate that, though. As is you kind of have to read through the code to know what is what in that display.

Reply:
I wonder how to do a benchmark on Colab that involves rebuilding the library? Ideally we should stay in the same Colab session for a fair comparison, but I know very little about Colab 😛

Reply:
> I think the output could be better labeled to indicate that, though. As is you kind of have to read through the code to know what is what in that display.

I would like to open a PR for that.

Reply:
> I would like to open a PR for that.

Sounds good to me. I think allowing `kwargs` in addition to `args` would also be useful.

Reply:
Thanks Gregory! That makes sense. So maybe we should be bumping `n` to a larger value?